src/ert/run_models/base_run_model.py

```diff
@@ -373,18 +421,159 @@ def checkHaveSufficientRealizations(
+            # Calculate completed realizations
+            current_iter = max(list(self._iter_snapshot.keys()))
+            done_reals = 0
+            all_reals = self._iter_snapshot[current_iter].reals
+            if not all_reals:
+                # Empty ensemble or all realizations deactivated
+                return 1.0
+            for real in all_reals.values():
+                if real.status in [
+                    REALIZATION_STATE_FINISHED,
+                    REALIZATION_STATE_FAILED,
+                ]:
+                    done_reals += 1
+            real_progress = float(done_reals) / len(all_reals)
+
+            return (
+                (current_iter + real_progress) / self.phaseCount()
+                if self.phaseCount() != 1
+                else real_progress
+            )
+
+    def send_end_event(self) -> None:
+        self.send_event(
+            EndEvent(
+                failed=self.hasRunFailed(),
+                failed_msg=self.getFailMessage(),
+            )
+        )
+
+    def send_snapshot_event(self, event: CloudEvent) -> None:
+        if event["type"] == EVTYPE_EE_SNAPSHOT:
+            iter_ = event.data["iter"]
+            snapshot = Snapshot(event.data)
+            self._iter_snapshot[iter_] = snapshot
+            self.send_event(
+                FullSnapshotEvent(
+                    phase_name=self.getPhaseName(),
+                    current_phase=self.currentPhase(),
+                    total_phases=self.phaseCount(),
+                    indeterminate=self.isIndeterminate(),
+                    progress=self._progress(),
+                    iteration=iter_,
+                    snapshot=copy.deepcopy(snapshot),
+                )
+            )
+        elif event["type"] == EVTYPE_EE_SNAPSHOT_UPDATE:
+            iter_ = event.data["iter"]
+            if iter_ not in self._iter_snapshot:
+                raise OutOfOrderSnapshotUpdateException(
+                    f"got {EVTYPE_EE_SNAPSHOT_UPDATE} without having stored "
+                    f"snapshot for iter {iter_}"
+                )
+            partial = PartialSnapshot(self._iter_snapshot[iter_]).from_cloudevent(event)
+            self._iter_snapshot[iter_].merge_event(partial)
+            self.send_event(
+                SnapshotUpdateEvent(
+                    phase_name=self.getPhaseName(),
+                    current_phase=self.currentPhase(),
+                    total_phases=self.phaseCount(),
+                    indeterminate=self.isIndeterminate(),
+                    progress=self._progress(),
+                    iteration=iter_,
+                    partial_snapshot=partial,
+                )
+            )
+
     def run_ensemble_evaluator(
         self, run_context: RunContext, ee_config: EvaluatorServerConfig
     ) -> List[int]:
+        if not self._end_queue.empty():
+            event_logger.debug("Run model canceled - pre evaluation")
+            self._end_queue.get()
+            return []
```

Comment:
Should this be a separate function we call periodically?

Reply:
Probably at a later point yes, we want it also in the analysis update :thinking: